// @wealthfolio/addon-sdk/dist/chunk-ISD4MMHY.js
// src/permissions.ts
var PERMISSION_CATEGORIES = [
  {
    id: "accounts",
    name: "Account Management",
    description: "Access to account information and settings",
    functions: ["getAll", "create"],
    riskLevel: "high"
  },
  {
    id: "portfolio",
    name: "Portfolio Data",
    description: "Access to holdings, portfolio performance, and account valuations",
    functions: [
      "getHoldings",
      "getHolding",
      "update",
      "recalculate",
      "getIncomeSummary",
      "getHistoricalValuations",
      "getLatestValuations"
    ],
    riskLevel: "high"
  },
  {
    id: "activities",
    name: "Transaction History",
    description: "Access to transaction records and activity management",
    functions: [
      "getAll",
      "search",
      "create",
      "update",
      "saveMany",
      "import",
      "checkImport",
      "getImportMapping",
      "saveImportMapping"
    ],
    riskLevel: "high"
  },
  {
    id: "market-data",
    name: "Market Data",
    description: "Access to market prices, quotes, and financial data",
    functions: ["searchTicker", "syncHistory", "sync", "getProviders", "fetchDividends"],
    riskLevel: "low"
  },
  {
    id: "assets",
    name: "Asset Management",
    description: "Access to asset profiles and data sources",
    functions: ["getProfile", "updateProfile", "updateQuoteMode"],
    riskLevel: "medium"
  },
  {
    id: "quotes",
    name: "Quote Management",
    description: "Access to price quotes and historical data",
    functions: ["update", "getHistory"],
    riskLevel: "low"
  },
  {
    id: "performance",
    name: "Performance Analytics",
    description: "Access to performance calculations and metrics",
    functions: ["calculateHistory", "calculateSummary", "calculateAccountsSimple"],
    riskLevel: "medium"
  },
  {
    id: "currency",
    name: "Exchange Rates",
    description: "Access to currency exchange rates and conversion data",
    functions: ["getAll", "update", "add"],
    riskLevel: "low"
  },
  {
    id: "financial-planning",
    name: "Financial Planning",
    description: "Access to financial goals and allocations",
    functions: [
      "getAll",
      "create",
      "update",
      "getFunding",
      "saveFunding",
      "updateAllocations",
      "getAllocations"
    ],
    riskLevel: "medium"
  },
  {
    id: "contribution-limits",
    name: "Contribution Limits",
    description: "Access to contribution limits and deposit calculations",
    functions: ["getAll", "create", "update", "calculateDeposits"],
    riskLevel: "medium"
  },
  {
    id: "settings",
    name: "Application Settings",
    description: "Access to application settings and configuration",
    functions: ["get", "update", "backupDatabase"],
    riskLevel: "medium"
  },
  {
    id: "files",
    name: "File Operations",
    description: "Access to file dialogs and file system operations",
    functions: ["openCsvDialog", "openSaveDialog"],
    riskLevel: "medium"
  },
  {
    id: "secrets",
    name: "Secrets Management",
    description: "Access to secure storage for addon secrets",
    functions: ["set", "get", "delete"],
    riskLevel: "high"
  },
  {
    id: "snapshots",
    name: "Snapshot Management",
    description: "Access to holdings snapshots for accounts with holdings tracking mode",
    functions: [
      "getAll",
      "getByDate",
      "save",
      "checkImport",
      "importSnapshots",
      "delete"
    ],
    riskLevel: "high"
  },
  {
    id: "events",
    name: "Event Listeners",
    description: "Access to application events and notifications",
    functions: [
      "onDropHover",
      "onDrop",
      "onDropCancelled",
      "onUpdateStart",
      "onUpdateComplete",
      "onUpdateError",
      "onSyncStart",
      "onSyncComplete"
    ],
    riskLevel: "low"
  },
  {
    id: "ui",
    name: "User Interface",
    description: "Access to modify navigation and add UI components",
    functions: ["sidebar.addItem", "router.add"],
    riskLevel: "low"
  }
];
function createFunctionPermission(name, isDeclared = false, isDetected = false, detectedAt) {
  return {
    name,
    isDeclared,
    isDetected,
    detectedAt: isDetected ? detectedAt || (/* @__PURE__ */ new Date()).toISOString() : void 0
  };
}
function getPermissionCategory(id) {
  return PERMISSION_CATEGORIES.find((category) => category.id === id);
}
function getPermissionCategoriesByRisk(riskLevel) {
  return PERMISSION_CATEGORIES.filter((category) => category.riskLevel === riskLevel);
}
function getFunctionRiskLevel(functionName) {
  const category = PERMISSION_CATEGORIES.find(
    (cat) => cat.functions.includes(functionName)
  );
  return category?.riskLevel;
}
function isPermissionRequired(functionName, categoryId) {
  const category = getPermissionCategory(categoryId);
  return category ? category.functions.includes(functionName) : false;
}
function getDeclaredFunctions(permission) {
  return permission.functions.filter((func) => func.isDeclared).map((func) => func.name);
}
function getDetectedFunctions(permission) {
  return permission.functions.filter((func) => func.isDetected).map((func) => func.name);
}
function getUndeclaredDetectedFunctions(permission) {
  return permission.functions.filter((func) => func.isDetected && !func.isDeclared).map((func) => func.name);
}
function hasUndeclaredDetectedFunctions(permission) {
  return permission.functions.some((func) => func.isDetected && !func.isDeclared);
}
function addDetectedFunction(permission, functionName, detectedAt) {
  const existingFunc = permission.functions.find((f) => f.name === functionName);
  if (existingFunc) {
    existingFunc.isDetected = true;
    existingFunc.detectedAt = detectedAt || (/* @__PURE__ */ new Date()).toISOString();
  } else {
    permission.functions.push(
      createFunctionPermission(functionName, false, true, detectedAt)
    );
  }
  return permission;
}
function markFunctionAsDeclared(permission, functionName) {
  const existingFunc = permission.functions.find((f) => f.name === functionName);
  if (existingFunc) {
    existingFunc.isDeclared = true;
  } else {
    permission.functions.push(createFunctionPermission(functionName, true, false));
  }
  return permission;
}

export {
  PERMISSION_CATEGORIES,
  createFunctionPermission,
  getPermissionCategory,
  getPermissionCategoriesByRisk,
  getFunctionRiskLevel,
  isPermissionRequired,
  getDeclaredFunctions,
  getDetectedFunctions,
  getUndeclaredDetectedFunctions,
  hasUndeclaredDetectedFunctions,
  addDetectedFunction,
  markFunctionAsDeclared
};
