// @wealthfolio/addon-sdk/dist/index.js
import {
  PERMISSION_CATEGORIES,
  getFunctionRiskLevel,
  getPermissionCategoriesByRisk,
  getPermissionCategory,
  isPermissionRequired
} from "./chunk-ISD4MMHY.js";

// src/query-keys.ts
var QueryKeys = {
  // Account related keys
  ACCOUNTS: "accounts",
  ACCOUNTS_SUMMARY: "accounts_summary",
  // Activity related keys
  ACTIVITY_DATA: "activity-data",
  ACTIVITIES: "activities",
  // Portfolio related keys
  HOLDINGS: "holdings",
  HOLDING: "holding",
  INCOME_SUMMARY: "incomeSummary",
  PORTFOLIO_SUMMARY: "portfolioSummary",
  QUOTE_HISTORY: "quoteHistory",
  // Goals related keys
  GOALS: "goals",
  GOALS_ALLOCATIONS: "goals_allocations",
  // Settings related keys
  SETTINGS: "settings",
  EXCHANGE_RATES: "exchangeRates",
  // New keys for exchange rates
  EXCHANGE_RATE_SYMBOLS: "exchange_rate_symbols",
  QUOTE: "quote",
  CONTRIBUTION_LIMITS: "contributionLimits",
  CONTRIBUTION_LIMIT_PROGRESS: "contributionLimitProgress",
  ASSET_DATA: "asset_data",
  IMPORT_MAPPING: "import_mapping",
  PERFORMANCE_SUMMARY: "performanceSummary",
  PERFORMANCE_HISTORY: "performanceHistory",
  HISTORY_VALUATION: "historyValuation",
  // Helper function to create account-specific keys
  valuationHistory: (id) => [QueryKeys.HISTORY_VALUATION, id],
  // Account simple performance
  ACCOUNTS_SIMPLE_PERFORMANCE: "accountsSimplePerformance",
  accountsSimplePerformance: (accountIds) => [
    QueryKeys.ACCOUNTS_SIMPLE_PERFORMANCE,
    [...accountIds].sort().join(",") || "none"
  ],
  // Market Data Providers
  MARKET_DATA_PROVIDERS: "marketDataProviders",
  MARKET_DATA_PROVIDER_SETTINGS: "marketDataProviderSettings",
  transactions: "transactions",
  latestValuations: "latest-valuations",
  // Market Data
  symbolSearch: "symbol-search",
  ASSET_HISTORY: "asset-history",
  // Addons
  INSTALLED_ADDONS: "installedAddons",
  ADDON_STORE_LISTINGS: "addonStoreListings",
  ADDON_AUTO_UPDATE_CHECK: "addonAutoUpdateCheck",
  SNAPSHOTS: "snapshots",
  snapshots: (accountId) => [QueryKeys.SNAPSHOTS, accountId],
  secrets: {
    apiKey: (providerId) => ["secrets", "apiKey", providerId]
  }
};

// src/manifest.ts
function isInstalledManifest(manifest) {
  return !!(manifest.installedAt && manifest.main !== void 0 && manifest.enabled !== void 0);
}

// package.json
var package_default = {
  name: "@wealthfolio/addon-sdk",
  version: "3.3.0",
  type: "module",
  description: "TypeScript SDK for building Wealthfolio addons with enhanced functionality and type safety",
  main: "dist/index.js",
  types: "dist/src/index.d.ts",
  exports: {
    ".": {
      import: "./dist/index.js",
      types: "./dist/src/index.d.ts"
    },
    "./types": {
      import: "./dist/types.js",
      types: "./dist/src/types.d.ts"
    },
    "./permissions": {
      import: "./dist/permissions.js",
      types: "./dist/src/permissions.d.ts"
    }
  },
  files: [
    "dist",
    "README.md",
    "CHANGELOG.md"
  ],
  keywords: [
    "wealthfolio",
    "addon",
    "plugin",
    "sdk",
    "typescript",
    "financial",
    "portfolio"
  ],
  author: "Wealthfolio Team",
  license: "MIT",
  homepage: "https://wealthfolio.app/addons",
  repository: {
    type: "git",
    url: "https://github.com/afadil/wealthfolio.git",
    directory: "packages/addon-sdk"
  },
  bugs: {
    url: "https://github.com/afadil/wealthfolio/issues"
  },
  scripts: {
    build: "tsup && pnpm run build:types",
    dev: "tsup --watch",
    clean: "rm -rf dist",
    lint: "eslint .",
    "lint:fix": "eslint . --fix",
    "lint:quiet": "eslint . --quiet",
    format: "prettier --write .",
    "format:check": "prettier --check .",
    "type-check": "tsc --noEmit",
    "build:types": "tsc -p tsconfig.json",
    prepack: "npm run build"
  },
  peerDependencies: {
    react: "^19.2.4",
    "react-dom": "^19.2.4"
  },
  devDependencies: {
    "@tanstack/react-query": "^5.90.20",
    "@types/react": "^19.2.13",
    "@types/react-dom": "^19.2.3",
    tsup: "^8.5.1",
    typescript: "^5.9.3"
  },
  engines: {
    node: ">=20.0.0"
  }
};

// src/version.ts
var SDK_VERSION = package_default.version;

// src/utils.ts
function validateManifest(manifest) {
  const errors = [];
  const warnings = [];
  if (!manifest.id) {
    errors.push("Addon ID is required");
  } else if (!/^[a-z0-9-]+$/.test(manifest.id)) {
    errors.push("Addon ID must contain only lowercase letters, numbers, and hyphens");
  }
  if (!manifest.name) {
    errors.push("Addon name is required");
  }
  if (!manifest.version) {
    errors.push("Addon version is required");
  } else if (!/^\d+\.\d+\.\d+/.test(manifest.version)) {
    warnings.push("Version should follow semantic versioning (e.g., 1.0.0)");
  }
  if (!manifest.description) {
    warnings.push("Description is recommended for better discoverability");
  }
  if (!manifest.author) {
    warnings.push("Author information is recommended");
  }
  if (!manifest.main) {
    warnings.push('Main entry point not specified, defaulting to "addon.js"');
  }
  if (manifest.permissions) {
    manifest.permissions.forEach((permission, index) => {
      if (!permission.category) {
        errors.push(`Permission ${index}: category is required`);
      }
      if (!permission.functions || permission.functions.length === 0) {
        errors.push(`Permission ${index}: at least one function must be specified`);
      }
      if (!permission.purpose) {
        warnings.push(`Permission ${index}: purpose explanation is recommended`);
      }
    });
  }
  return {
    valid: errors.length === 0,
    errors,
    warnings
  };
}
function isCompatibleVersion(addonSdkVersion, currentSdkVersion = SDK_VERSION) {
  if (!addonSdkVersion) return true;
  const [addonMajor, addonMinor] = addonSdkVersion.split(".").map(Number);
  const [currentMajor, currentMinor] = currentSdkVersion.split(".").map(Number);
  return addonMajor === currentMajor && addonMinor <= currentMinor;
}
function formatAddonSize(bytes) {
  const sizes = ["B", "KB", "MB", "GB"];
  if (bytes === 0) return "0 B";
  const i = Math.floor(Math.log(bytes) / Math.log(1024));
  const size = bytes / Math.pow(1024, i);
  return `${size.toFixed(i === 0 ? 0 : 1)} ${sizes[i]}`;
}
function generateAddonId(name) {
  return name.toLowerCase().replace(/[^a-z0-9\s-]/g, "").replace(/\s+/g, "-").replace(/-+/g, "-").replace(/^-|-$/g, "");
}
function isAddonManifest(obj) {
  return typeof obj === "object" && obj !== null && typeof obj.id === "string" && typeof obj.name === "string" && typeof obj.version === "string";
}

// src/goal-progress.ts
function isParticipatingGoal(goal) {
  if (goal.statusLifecycle) {
    return goal.statusLifecycle === "active";
  }
  return true;
}
function getFiniteAmount(value) {
  if (value === null || value === void 0 || value === "") {
    return void 0;
  }
  const amount = typeof value === "number" ? value : Number(value);
  return Number.isFinite(amount) ? amount : void 0;
}
function toFiniteAmount(value) {
  return getFiniteAmount(value) ?? 0;
}
function calculateGoalProgress(accountsValuations, goals, allocations) {
  if (!goals) {
    return [];
  }
  const baseCurrency = accountsValuations?.[0]?.baseCurrency ?? "USD";
  const accountValueMap = /* @__PURE__ */ new Map();
  accountsValuations?.forEach((account) => {
    const valueInBaseCurrency = (account.totalValue ?? 0) * (account.fxRateToBase ?? 1);
    accountValueMap.set(account.accountId, valueInBaseCurrency);
  });
  const allocationsByGoal = /* @__PURE__ */ new Map();
  allocations?.forEach((alloc) => {
    const existing = allocationsByGoal.get(alloc.goalId) ?? [];
    allocationsByGoal.set(alloc.goalId, [...existing, alloc]);
  });
  const sortedGoals = [...goals].filter(isParticipatingGoal).sort(
    (a, b) => toFiniteAmount(a.summaryTargetAmount ?? a.targetAmount) - toFiniteAmount(b.summaryTargetAmount ?? b.targetAmount)
  );
  return sortedGoals.map((goal) => {
    const goalAllocations = allocationsByGoal.get(goal.id) ?? [];
    const targetAmount = toFiniteAmount(goal.summaryTargetAmount ?? goal.targetAmount);
    const totalAllocatedValue = goalAllocations.reduce((total, allocation) => {
      const accountValueInBase = accountValueMap.get(allocation.accountId) ?? 0;
      return total + accountValueInBase * allocation.sharePercent / 100;
    }, 0);
    const currentValue = getFiniteAmount(goal.summaryCurrentValue) ?? totalAllocatedValue;
    const progress = getFiniteAmount(goal.summaryProgress) ?? (targetAmount > 0 ? currentValue / targetAmount : 0);
    return {
      goalId: goal.id,
      name: goal.title,
      targetValue: targetAmount,
      currentValue,
      progress,
      currency: goal.currency ?? baseCurrency,
      statusHealth: goal.statusHealth,
      targetDate: goal.targetDate
    };
  });
}

// src/index.ts
var ReactVersion = "19.1.1";
var hostGlobals = window;
var React = hostGlobals.React;
var ReactDOM = hostGlobals.ReactDOM;
export {
  PERMISSION_CATEGORIES,
  QueryKeys,
  React,
  ReactDOM,
  ReactVersion,
  SDK_VERSION,
  calculateGoalProgress,
  formatAddonSize,
  generateAddonId,
  getFunctionRiskLevel,
  getPermissionCategoriesByRisk,
  getPermissionCategory,
  isAddonManifest,
  isCompatibleVersion,
  isInstalledManifest,
  isPermissionRequired,
  validateManifest
};
/**
 * @wealthfolio/addon-sdk
 *
 * TypeScript SDK for building Wealthfolio addons with enhanced functionality,
 * type safety, and comprehensive permission management.
 *
 * @version 1.0.0
 * @author Wealthfolio Team
 * @license MIT
 */
